### PR TITLE
Add threshold expert bot

### PR DESCRIPTION
**Pull request by creideiki**
Bot for only propagating messages if a certain number of similar messages arrive in a certain time.

### CODE REVIEW

File: intelmq/bots/experts/threshold/expert.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                                        filter_type=self.filter_type)
+            old_count = int(self.cache.get(message_hash) or 0)
+            self.logger.debug('Message %s has been seen %i times before.',
+                              message_hash, old_count)
+            # Use Redis "set" instead of "incr" to reset the timeout
+            # every time
+            self.cache.set(message_hash, str(old_count + 1))
+            if old_count + 1 == self.threshold:
+                self.logger.debug('Threshold reached, forwarding message.')
+                message.update(self.add_keys)
```
Comment (wagner-certat) [inline intelmq/bots/experts/threshold/expert.py:93]: I suggest to also add the used threshold, maybe as `extra.count` (as this name is already in use by some other bots).

**Review by wagner-certat — CHANGES_REQUESTED** [review]
Another big thanks for providing this contribution!

File: intelmq/bots/experts/threshold/expert.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+"""Threshold value expert bot
+
+SPDX-FileCopyrightText: 2020 Linköping University <https://liu.se/>
+SPDX-License-Identifier: AGPL-3.0-or-later
+
+Given a stream of messages, this bot will let through only the single
+one that makes the cound of similar messages go above a threshold
+value.
+
+This bot is not multiprocessing safe. Do not run more than one
```
[the remark this answers is not in the document]
Comment (creideiki) [reply intelmq/bots/experts/threshold/expert.py:11]: Fixed.

File: docs/Bots.md [hunk 3]
```diff
@@ -2491,6 +2492,33 @@ For brevity, "type" means `classification.type` and "taxonomy" means `classifica
...
+### Threshold
+
+#### Information:
+
+* **Cache parameters** (see in section [common parameters](#common-parameters))
+* `name`: threshold
+* `lookup`: redis cache
+* `public`: no
+* `cache (redis db)`: 11
+* `description`: Check if the number of similar messages during a specified time interval exceeds a set value.
```
[the remark this answers is not in the document]
Comment (creideiki) [reply docs/Bots.md:2504]: You are correct. I really wanted to have a rolling window so the bot could look at exactly the last time window, but that would require keeping large amounts of state. This solution is good enough for my use case, but I have no idea how to succinctly describe it (other than restating what the code does, as in the description of the `timeout` parameter). If you have a suggestion for a better wording, I'll happily accept it.

File: intelmq/bots/experts/threshold/expert.py (hunk 1 above)
Comment (creideiki) [reply intelmq/bots/experts/threshold/expert.py:97]: Done.

File: docs/Bots.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (wagner-certat) [reply docs/Bots.md:2504]: As I found no better wording too, I now simply explained the algorithm in the documentation.

**Comment by wagner-certat** [comment]
Thanks again for your contribution!

**Pull request merged**